### ISSUE TITLE
[ENH] add print timestap of 'elapsed time' to each workflow

### ISSUE THREAD

**Issue by marcobarilari** [cpp-lln-lab/CPP_SPM#228]
You think / need this code to do **keep track of time if I run long data crunch session**.

That means that we can improve the code or the documentation.

Tell us more about it below.

It can be a separate func with the following code

```matlab
tic

[...]

fprintf('\n\n\n********  Done   ********\n\n');
fprintf('elapsed time is: ')
t=toc;
disp(datestr(datenum(0,0,0,0,0,t),'HH:MM:SS'));
fprintf('\n*************************\n\n');
```

that will print in the command window in `HH:MM:SS` eg

```
********  Done   ********

elapsed time is: 00:08:25

*************************
```

**Comment by marcobarilari** [comment]
it can also have a smile :)

```
fprintf('\n\n\n******** Done :) ********\n\n');
fprintf('elapsed time is: ')
t=toc;
disp(datestr(datenum(0,0,0,0,0,t),'HH:MM:SS'));
fprintf('\n*************************\n\n');
```

**Comment by Remi-Gau** [comment]
do you know if that can be made into a function?

**Comment by marcobarilari** [comment]
yes,

in the workflow func:

```
elapsedTime('start');


WaitSecs(5);


elapsedTime('stop');
```

in the func:

```
function elapsedTime(input)

switch input
    
    case 'start' 
        
        tic;
    
    
    case 'stop' 
        
        fprintf('\n\n\n******** Done :) ********\n\n');
        fprintf('elapsed time is: ');
        t=toc;
        disp(datestr(datenum(0,0,0,0,0,t),'HH:MM:SS'));
        fprintf('\n*************************\n\n');
        
end
```

**Comment by Remi-Gau** [comment]
cool let's open a PR and add this to our workflows.

I would actually add it in `setUpWorkflow` and `saveAndRunWorkflow`, this way the function will always be called in all our workflows at once.

https://github.com/cpp-lln-lab/CPP_SPM/blob/dev/src/workflows/setUpWorkflow.m

https://github.com/cpp-lln-lab/CPP_SPM/blob/dev/src/workflows/saveAndRunWorkflow.m

**Comment by marcobarilari** [comment]
The idea to add it to a small nb of functions that are called by all the workflows would be the best but I don't think that adding it to `setUpWorkflow` and `saveAndRunWorkflow` would do it. 

For example, in `bidsCopyRawFolder` the funcs above are not there and in `bidsSTC` `saveAndRunWorkflow` is called for every subject meaning that you have not a global elapsed time output at the end of the entire dataset STC analysis. 

I guess that for the moment the option that covers all the workflows for the global duration is to add `elapsedTime('start');` and `elapsedTime('stop');` in each script. 

But yeah your idea would save time so if something else comes up to your mind LMK

**Comment by Remi-Gau** [comment]
> For example, in `bidsCopyRawFolder` the funcs above are not there 

I think it is possible to add them there, no?


> and in `bidsSTC` `saveAndRunWorkflow` is called for every subject meaning that you have not a global elapsed time output at the end of the entire dataset STC analysis.

True but that's a bad "side effect" of the fact that the "API" for our `setBath*` is not standardized enough.

See #184 and #82.

If all the the `setBatch` functions took `matlabbatch` as input argument then we could have a single call to `saveAndRunWorkflow` per workflow.

I have been pospotning this change because it means changing a lot of stuff in many different places.
 
> I guess that for the moment the option that covers all the workflows for the global duration is to add `elapsedTime('start');` and `elapsedTime('stop');` in each script.

You mean in the scripts them selves and not in any function?

**Comment by Remi-Gau** [comment]
Actually, the previous post makes me realize that the way our workflow function function might indeed not work if we want to have an overall elapsed time for a whole workflow and not for every subject.

Also what's the issue with getting the time elapsed for each subject ? Why would you want it for the whole workflow? 
What would be nice would be to get the time elapsed for a given subject subject and then an ETA to have all the subject to finish (using the mean elapsed time for each of the previous subjects multiplied by the number of subjects remaining).

**Comment by marcobarilari** [comment]
indeed is more complicated than I thought. 

> You mean in the scripts themselves and not in any `function?`

I mean in each workflow

> Also what's the issue with getting the time elapsed for each subject? Why would you want it for the whole workflow?

Thinking twice, it would be informative to have both the elapsed time for the entire workflow and for each subj, might be helpful to plan analysis on the cluster, for example, given time constrain.

> What would be nice would be to get the time elapsed for a given subject subject and then an ETA to have all the subject to finish (using the mean elapsed time for each of the previous subjects multiplied by the number of subjects remaining).

this is also a good feature

Ok, let's plan this function to have at least 3 outcomes:

(1) elapsed time for a subject
(2) elapsed time for the entire workflow
(3) ETA for the entire workflow given the first subj (I don't if it is needed to update this on each subj)
(4) - optional - global elapsed time at the level of the task batch script 

what do you think?

**Comment by Remi-Gau** [comment]
yes sounds good!!

**Comment by Remi-Gau** [comment]
> > You mean in the scripts themselves and not in any `function?`
> 
> I mean in each workflow

Not a big fan of duplicating the same code in EACH workflow. So I would do that only if it is REALLY necessary

**Comment by Remi-Gau** [comment]
Related issue: what is the best way to save our "logs": one per workflow across several subjects or one per workflow / per subject (as we do now).

**Comment by Remi-Gau** [comment]
closed by #229
